soweego/linker/evaluate.py

```diff
@@ -75,58 +134,93 @@ def _compute_performance(test_index, predictions, test_vectors_size):
+        raise NotImplementedError(f'Grid search for {classifier} is not supported')
+
+    result = defaultdict(list)
+
+    dataset, positive_samples_index = train.build_dataset(
+        'training', catalog, entity, dir_io)
+    model = workflow.init_model(classifier, **kwargs).kernel
+
+    inner_k_fold, target = utils.prepare_stratified_k_fold(
+        k, dataset, positive_samples_index)
+    outer_k_fold = StratifiedKFold(n_splits=k, shuffle=True, random_state=1269)
+    grid_search = GridSearchCV(
+        model, param_grid, scoring=scoring, n_jobs=-1, cv=inner_k_fold, verbose=2)
+    dataset = dataset.to_numpy()
+
+    for train_index, test_index in outer_k_fold.split(dataset, target):
+        # Run grid search
+        grid_search.fit(dataset[train_index], target[train_index])
+        # Grid search best score is the train score
+        result[f'train_{scoring}'].append(grid_search.best_score_)
+        # Let grid search compute the test score
+        test_score = grid_search.score(dataset[test_index], target[test_index])
+        result[f'test_{scoring}'].append(test_score)
+        best_model = grid_search.best_estimator_
+        result['best_models'].append(best_model)
+
+    return result
+
+
+def average_k_fold(classifier, catalog, entity, k, dir_io, **kwargs):
     predictions, precisions, recalls, fscores = None, [], [], []
     dataset, positive_samples_index = train.build_dataset(
         'training', catalog, entity, dir_io)
-    k_fold = StratifiedKFold(n_splits=k, shuffle=True)
-    # scikit's stratified k-fold no longer supports multi-label data representation.
-    # It expects a binary array instead, so build it based on the positive samples index
-    binary_target_variables = dataset.index.map(
-        lambda x: 1 if x in positive_samples_index else 0)
+    k_fold, binary_target_variables = utils.prepare_stratified_k_fold(
+        k, dataset, positive_samples_index)
 
     for train_index, test_index in k_fold.split(dataset, binary_target_variables):
         training, test = dataset.iloc[train_index], dataset.iloc[test_index]
 
-        model = workflow.init_model(classifier, binarize, training.shape[1])
+        model = _initialize(classifier, dataset, kwargs)
         model.fit(training, positive_samples_index & training.index)
-        
+
         preds = model.predict(test)
-        
+
         p, r, f, _ = _compute_performance(
             positive_samples_index & test.index, preds, len(test))
 
         if predictions is None:
             predictions = preds
         else:
             predictions |= preds
-        
+
         precisions.append(p)
         recalls.append(r)
         fscores.append(f)
 
     return predictions, mean(precisions), std(precisions), mean(recalls), std(recalls), mean(fscores), std(fscores)
 
 
-def single_k_fold(classifier, catalog, entity, binarize, dir_io, k=5):
+def _initialize(classifier, dataset, kwargs):
+    if classifier is constants.SINGLE_LAYER_PERCEPTRON:
```

Comment:
Also here, after merging with master the `mlp` classifier should also be considered 

```suggestion
    if classifier in (constants.SINGLE_LAYER_PERCEPTRON, 
                      constants.MULTILAYER_CLASSIFIER):
```